pytorch_lightning/utilities/warnings.py

```diff
@@ -11,17 +11,39 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from pytorch_lightning.utilities.distributed import rank_zero_deprecation, rank_zero_warn
+import warnings
+from functools import partial
+
+from pytorch_lightning.utilities.distributed import rank_zero_only
+
+
+def _warn(*args, stacklevel: int = 2, **kwargs):
+    warnings.warn(*args, stacklevel=stacklevel, **kwargs)
+
+
+@rank_zero_only
+def rank_zero_warn(*args, stacklevel: int = 4, **kwargs):
```

Comment:
why are we splitting these apart?
can you keep them in the same file?

Reply:
1. Because it only makes sense that warnings-related stuff goes into the warnings.py files
2. This new structure reduces the circular conflicts (although there are still some).

If I keep the previous structure, then `distributed.py` would have needed to import `LightningDistributedWarning` but the `warning_cache` already imports `rank_zero_{warn,deprecation}` from distributed.py.

That is unless you think `LightningDistributedWarning` should be defined in `distributed.py`.

Now the only conflicts is that `warnings.py` needs `rank_zero_only` from `distributed.py` and `register_ddp_comm_hook` needs to warn.

IMO warnings are good where they are in this PR and the logging constructs should be moved out of `distributed.py`

Reply:
mind add a comment about this reasoning to code bc quite sure the next refactor will try to get them together...